src/metrics.rs

```diff
@@ -1,30 +1,137 @@
 
-    #[metric(describe = "Latency for `engine_getPayloadV3`")]
+    #[metric(describe = "Number of client `engine_newPayloadV3` responses", labels = ["code"])]
+    #[allow(dead_code)]
+    pub new_payload_v3_response_count: Counter,
+
+    #[metric(describe = "Latency for client `engine_getPayloadV3` call")]
+    #[allow(dead_code)]
     pub get_payload_v3: Histogram,
 
-    #[metric(describe = "Latency for `engine_forkChoiceUpdatedV3`")]
+    #[metric(describe = "Number of client `engine_getPayloadV3` responses", labels = ["code"])]
+    #[allow(dead_code)]
+    pub get_payload_v3_response_count: Counter,
+
+    #[metric(describe = "Latency for client `engine_forkChoiceUpdatedV3` call")]
+    #[allow(dead_code)]
     pub fork_choice_updated_v3: Histogram,
 
-    #[metric(describe = "Count of blocks created by the builder")]
-    pub blocks_created_by_builder: Counter,
+    #[metric(describe = "Number of client `engine_forkChoiceUpdatedV3` responses", labels = ["code"])]
+    #[allow(dead_code)]
+    pub fork_choice_updated_v3_response_count: Counter,
+}
+
+impl ServerMetrics {
+    pub fn increment_blocks_created(&self, source: &PayloadSource) {
+        counter!("rpc.blocks_created", "source" => source.to_string()).increment(1);
+    }
+
+    pub fn record_builder_forwarded_call(&self, latency: Duration, method: String) {
+        histogram!("rpc.builder_forwarded_call", "method" => method).record(latency.as_secs_f64());
+    }
+
+    pub fn increment_builder_rpc_response_count(
+        &self,
+        http_status_code: String,
+        rpc_status_code: Option<String>,
+        method: String,
+    ) {
+        counter!("rpc.builder_response_count",
+            "http_status_code" => http_status_code,
+            "rpc_status_code" => rpc_status_code.unwrap_or("".to_string()),
+            "method" => method,
+        )
+        .increment(1);
+    }
+
+    pub fn record_l2_forwarded_call(&self, latency: Duration, method: String) {
+        histogram!("rpc.l2_forwarded_call", "method" => method).record(latency.as_secs_f64());
+    }
+
+    pub fn increment_l2_rpc_response_count(
+        &self,
+        http_status_code: String,
+        rpc_status_code: Option<String>,
+        method: String,
+    ) {
+        counter!("rpc.l2_response_count",
+            "http_status_code" => http_status_code,
+            "rpc_status_code" => rpc_status_code.unwrap_or("".to_string()),
+            "method" => method,
+        )
+        .increment(1);
+    }
+}
+
+impl ClientMetrics {
+    pub fn new(source: &PayloadSource) -> Self {
+        Self {
+            new_payload_v3: histogram!("rpc.new_payload_v3", "target" => source.to_string()),
+            get_payload_v3: histogram!("rpc.get_payload_v3", "target" => source.to_string()),
+            fork_choice_updated_v3: histogram!("rpc.fork_choice_updated_v3", "target" => source.to_string()),
+            new_payload_v3_response_count: counter!("rpc.new_payload_v3_response_count"),
```

Comment:
Doesnt histogram also cover counts?

Reply:
its a bit hard to add the status code attribute without a separate metric since the target would be overwritten so I opted for a separate count metric